Use cachable flash address space while reading on infineon xmc4xxx

talih0 opened the pull request:
The infineon xmc4xxx series has two ways to access flash: One is the
cachable address space at 0x8000000 which can use pre-fetched data to
reduce flash access latency; The second is uncached space at 0xc000000
which is mainly used for write and erase operations.
The two address spaces point to the physical flash.
    
This patch adds support for using the cachable read access by setting
it as the virtual address in the linker script for the rom section.

talih0 commented:
I've split up the pull request to simplify review and fixed some compliance checks.

parthitce commented on dts/arm/infineon/xmc4500.dtsi:
```diff
@@ -21,5 +21,5 @@
 
 &flash0 {
-	reg = <0xc000000 DT_SIZE_M(1)>;
+	reg = <0xc000000 DT_SIZE_M(1) 0x8000000 DT_SIZE_M(1)>;
```
So the idea here is to use cached address space always? If so, I need to read back the TRM about flushing the cache during abnormal scenarios like reset.

talih0 replied:
> 

That's how it's done in the Infineon linker script. I didn't see in the manual (or example code startup code) about handling these cases. 

There can be some cache coherency problems when writing/erasing cache and then reading from cached address space..  (but we don't have a flash driver yet):

![Screenshot_2022-07-03_10-12-48](https://user-images.githubusercontent.com/17433307/177044052-c3d3fef5-37d7-4eb4-8ea0-b295bf6955fc.png)

talih0 replied:
> So the idea here is to use cached address space always?

The user can also go back to the previous version (using uncached address space) by setting:
reg = <0xc000000 DT_SIZE_M(1) 0xc000000 DT_SIZE_M(1)>;

parthitce commented on include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -207,8 +208,9 @@ SECTIONS
 
 	__rodata_region_end = .;
-	MPU_ALIGN(__rodata_region_end -__rom_region_start);
+	MPU_ALIGN(__rodata_region_end - ADDR(rom_start));
```
IMO these changes split into another PR to keep the arch changes independent of xmc?

talih0 replied:
Yes, I think that's a good idea. For now, I'd like to keep the XMC4xxx changes here to add some context to the linker changes. I'll split up the PR in a few days.

parthitce replied:
@talih0 possible to split this up?

talih0 replied:
> @talih0 possible to split this up?

I've put these changes in a separate PR.

talih0 committed:
soc: arm: infineon_xmc: Add support for using cachable flash address

The infineon xmc4xxx series has two ways to access flash: one is the
cachable address space at 0x8000000 which can use pre-fetched data to
reduce flash access latency; the second is uncached space at 0xc000000
which is mainly used for write and erase operations.

This patch adds support for using the cachable read access by setting
it as the virtual address in the linker script for the rom section.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 commented:
I found some more linker snippets that need an ALIGN(4) to sync up the vma and lma addresses. I'll submit a separate PR for those changes. 
I'll set this PR to draft until those changes are merged.
